### ISSUE TITLE
for GNN

### ISSUE THREAD

**Issue by aozorahime** [median-research-group/LibMTL#24]
Hi, I know it's kinda OOT but I am curious whether I can apply multi-task learning to graph neural network. What I learn from HPS, we shall share the encoder/decoder across the layer. I am curious should I create an encoder on top of the graph layer? Kinda stuck in this experiment, any suggestion would be helpful. Thanks

**Comment by Baijiong-Lin** [comment]
I know very little about the graph neural network but I think maybe you can try to define the encoder in HPS as some graph layers to extract some shared features among tasks.

**Comment by aozorahime** [comment]
I see. How about the attention layer on top of graph, I shall define the encoder after that, right? Do you have an MTL  example (either code or reference) for text classification?

**Comment by Baijiong-Lin** [comment]
We have no such example using GNN in LibMTL now. But there are many works about multi-task text classification and I think it is easy to find them on Google or Github. You can refer to how a multi-task GNN is built.

**Comment by aozorahime** [comment]
yes, thank you for your suggestion

**Comment by Baijiong-Lin** [comment]
Closed as no further updates.

**Comment by davodogster** [comment]
@aozorahime Hi I have been using PyTorch geometric and added my own code to make it work for multi-task learning. Let me know if you're interested.

@Baijiong-Lin Cool repo! I am interested in using loss weights such as RLW for my Loss.
Do you know how I can put the two together? It is important because my paramaters are not on the same scale so I want to weight the losses accordingly. Thanks! Sam

Will something like this work for my data?
![image](https://user-images.githubusercontent.com/46079516/223914150-5ef83f4e-5027-43a7-9b38-46e273b9ec93.png)

Here is my current code for how I obtain the losses:
![image](https://user-images.githubusercontent.com/46079516/223914346-92064f35-0ebd-470d-ad58-4d75e7cb1b7e.png)

**Comment by Baijiong-Lin** [comment]
@davodogster Maybe you can try it, I am not sure whether it can work in such a GNN case.

**Comment by aozorahime** [comment]
@davodogster Hi, cool you made it! Have you published it in your repo?

**Comment by davodogster** [comment]
@aozorahime No Sorry, I haven't published anything yet. How urgent is it?
Hopefully this can help for now:

![image](https://user-images.githubusercontent.com/46079516/224197066-2caca0a0-1495-4645-8740-954854532870.png)
the "3" represents 3 different regression tasks. 

![image](https://user-images.githubusercontent.com/46079516/224197219-e9e4b8c7-2f2c-4e49-be53-65774b967c8f.png)
![image](https://user-images.githubusercontent.com/46079516/224197429-841d35c7-cfe9-456d-8d1c-a880953e7d01.png)

PyG Dataset example:
![image](https://user-images.githubusercontent.com/46079516/224197548-bc18db8e-0831-4320-b386-595ad5d5f17f.png)

This gets passed to the PyG DatsetClass

![image](https://user-images.githubusercontent.com/46079516/224197651-426ab99c-ec90-4f3b-9aaa-a6f74aefc14f.png)

Dataset gets passed to the dataloader

test_loader = DataLoader(DBH_DS_test, batch_size=TRAIN_BS, shuffle = False, num_workers=25)

**Comment by davodogster** [comment]
> @davodogster Maybe you can try it, I am not sure whether it can work in such a GNN case.

@Baijiong-Lin Do you know what variable and shape is "task_num" in the example above? How do I modify it for three-task learning?

I think it might be able to work for my use case but I will need some tips :)

**Comment by aozorahime** [comment]
@davodogster Cool! so, the point is you use three different losses for the task right? MTL is for my thesis currently. Can we discuss this further later? or is it okay to discuss this here? :D

**Comment by davodogster** [comment]
@aozorahime yup you use three losses. It can be a mix of classification and regression losses depending on the tasks/params. Then you either sum the losses or use a technique like weighting the losses / GradNorm etc. OR you can scale all features if regression, to be on the same scale e.g. mean 0, SD 1. Or minMax = 0-100. then I think you dont need to weight the loss.
I'm using this one

https://github.com/ywatanabe1989/custom_losses_pytorch/blob/master/multi_task_loss.py

Yes we can discuss it more some time if needed - March is a really busy month for me this year but I can try to fit it in some time! :)
